Orange/widgets/utils/webview.py

```diff
@@ -92,31 +183,230 @@ def runJavaScript(self, javascript, resultCallback=None):
+    pyqt_type = type(obj).__mro__[-2]
+    if isinstance(obj, (list, np.ndarray)):
+        pyqt_type = 'QVariantList'
+    elif isinstance(obj, Mapping):
+        pyqt_type = 'QVariantMap'
+    elif isinstance(obj, QObject):
+        pyqt_type = QObject
+    else:
+        raise TypeError(
+            "Can't expose object of type {}. Too easy. Use "
+            "evalJS method instead.".format(type(obj)))
+
+    class _JSObject(QObject):
+        """ This class hopefully prevent options data from being
+        marshalled into a string-like dumb (JSON) object when
+        passed into JavaScript. Or at least relies on Qt to do it as
+        optimally as it knows to."""
+
+        # JS webchannel listens to this signal
+        objectChanged = pyqtSignal()
+
+        def set_object(self, obj):
+            self._obj = obj
+            self.objectChanged.emit()
+
+        @pyqtProperty(pyqt_type, notify=objectChanged)
+        def object(self):
+            return self._obj
+
+    jsobj = _JSObject(parent)
+    jsobj.set_object(obj)
+    return jsobj
+
+
+if HAVE_WEBKIT:
+
+    class WebviewWidget(WebKitView, _WebViewBase):
+        def evalJS(self, code):
+            return self.frame.evaluateJavaScript(code)
+
+        def onloadJS(self, code):
+            self.frame.loadFinished.connect(
+                lambda: WebviewWidget.evalJS(self, code))
+
+        def html(self):
+            return self.frame.toHtml()
+
+        def exposeObject(self, name, obj):
+            obj = _to_primitive_types(obj)
+            jsobj = _JSObject(self, obj)
+            self.frame.addToJavaScriptWindowObject('_' + name, jsobj)
+            WebviewWidget.evalJS(self, '''
+                window.{0} = window._{0}.object;
+                __fixupPythonObject({0});
+            '''.format(name))
+
+elif HAVE_WEBENGINE:
+
+    class WebviewWidget(WebEngineView, _WebViewBase):
+        class _NOTSET:
+            pass
+        _html = _NOTSET
+        _result = _NOTSET
+
+        # The alternative to QWebChannel.registeredObjects() that doesn't
+        # segfault or raise RuntimeErrors about objects being deleted
+        _registered_objects = None
+
+        # Tracks objects exposed in JS via exposeObject(). JS notifies once
+        # the object has indeed been exposed (i.e. the new object is available
+        # is JS) because the whole thing is async.
+        # This is needed to stall any evalJS() calls which may expect
+        # the objects being available (but could otherwise be executed before
+        # the objects are exposed in JS).
+        _to_be_exposed = None
+
+        def evalJS(self, code):
+            while not self._is_objects_exposed:
+                qApp.processEvents()
+            self.runJavaScript(code, lambda result: setattr(self, '_result', result))
+            while self._result is self._NOTSET:
+                qApp.processEvents()
+            result, self._result = self._result, self._NOTSET
+            return result
```

Comment:
At the very least this needs to guard against reentry. It also needs to document that it calls `qApp.processEvents`. Similar for all other methods that call `qApp.processEvents`